array_api_compat/common/_helpers.py

```diff
@@ -801,6 +803,262 @@ def size(x):
+        Otherwise, (None, preprocessed x)
+        """
+        if self.idx is _undef:
+            raise TypeError(
+                "Index has not been set.\n"
+                "Usage: either\n"
+                "    at(x, idx).set(value)\n"
+                "or\n"
+                "    at(x)[idx].set(value)\n"
+                "(same for all other methods)."
+            )
+        
+        x = self.x
+
+        if copy is False:
+            if not is_writeable_array(x) or is_dask_array(x):
+                raise ValueError("Cannot modify parameter in place")
+        elif copy is None:
+            copy = not is_writeable_array(x)
+        elif copy == "_force_false":
+            copy = False
+        elif copy is not True:
+            raise ValueError(f"Invalid value for copy: {copy!r}")
+
+        if is_jax_array(x):
+            # Use JAX's at[]
+            at_ = x.at[self.idx]
+            args = (y,) if y is not _undef else ()
+            return getattr(at_, at_op)(*args, **kwargs), None
+
+        # Emulate at[] behaviour for non-JAX arrays
+        if copy:
+            # FIXME We blindly expect the output of x.copy() to be always writeable.
+            # This holds true for read-only numpy arrays, but not necessarily for
+            # other backends.
+            xp = array_namespace(x)
+            x = xp.asarray(x, copy=True)
+
+        return None, x
+
+    def get(self, copy: bool | None = True, **kwargs):
+        """
+        Return x[idx]. In addition to plain __getitem__, this allows ensuring
+        that the output is either a copy or a view; it also allows passing
+        kwargs to the backend.
+        """
+        # __getitem__ with a fancy index always returns a copy.
+        # Avoid an unnecessary double copy.
+        # If copy is forced to False, raise.
+        if _is_fancy_index(self.idx):
+            if copy is False:
+                raise TypeError(
+                    "Indexing a numpy array with a fancy index always "
+                    "results in a copy"
+                )
+            # Skip copy inside _common, even if array is not writeable
+            copy = "_force_false"  # type: ignore
+
+        res, x = self._common("get", copy=copy, **kwargs)
+        if res is not None:
+            return res
+        return x[self.idx]
+
+    def set(self, y, /, **kwargs):
+        """x[idx] = y"""
+        res, x = self._common("set", y, **kwargs)
+        if res is not None:
+            return res
+        x[self.idx] = y
+        return x
+
+    def apply(self, ufunc, /, **kwargs):
+        """ufunc.at(x, idx)"""
+        if is_cupy_array(self.x) or is_torch_array(self.x) or is_dask_array(self.x):
+            # ufunc.at not implemented
+            return self.set(ufunc(self.x[self.idx]), **kwargs)
+
+        res, x = self._common("apply", ufunc, **kwargs)
+        if res is not None:
+            return res
+        ufunc.at(x, self.idx)
+        return x
```

Comment:
IMHO we should delete this whole method (full rationale: https://github.com/data-apis/array-api-compat/pull/205#issuecomment-2504865475)